third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,256 @@
+        threads_per_block * 2 <= threads_per_block_max) {
+      ip.batch /= 2;
+      threads_per_block *= 2;
+    } else if (
+        ip.batch % 3 == 0 && ip.batch / 3 >= batch_min &&
+        threads_per_block * 3 <= threads_per_block_max) {
+      ip.batch /= 3;
+      threads_per_block *= 3;
+    } else if (
+        ip.batch % 5 == 0 && ip.batch / 5 >= batch_min &&
+        threads_per_block * 5 <= threads_per_block_max) {
+      ip.batch /= 5;
+      threads_per_block *= 5;
+    } else {
+      break;
+    }
+  }
+
+  // Step-2, set InnerParams Iteration dim: gdimy. reg_per_thread is estimated
+  // from buffer size, then it is used to calculate threads_per_sm and gdimy.
+  // gdimy_max ensures each block processes at least 8 rows to
+  // reduce the workload of the final outer reduction.
+  int64_t reg_per_thread = getEstimatedRegisterUsage(ip.vect * ip.batch);
+  int64_t threads_per_sm = getThreadsPerSMGivenRegPerThread(reg_per_thread);
+  int64_t blocks_per_sm =
+      getBlocksPerSM(threads_per_sm, threads_per_block, dev_prop->warpSize);
+  ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+  const int64_t outer_iter_min = 8;
+  const int64_t gdimy_max = scheduler_utils::roundUpToN(
+      ceilDiv(outer_dim_numel, outer_iter_min), device_multiprocessor_count);
+  while (ip.gdimy > gdimy_max && blocks_per_sm > 1) {
+    blocks_per_sm -= 1;
+    ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+  }
+
+  // Step-3, set OuterParams Iteration dim: unroll_factor, bdimx, gdimy (already
+  // done)
+  op.unroll_factor = inner_dim_numel >= 4096 ? 4 : 2;
+  op.gdimy = ip.gdimy;
+  op.bdimx = scheduler_utils::roundUpPow2Or8(
+      ceilDiv(inner_dim_numel / op.unroll_factor, op.gdimy));
+
+  // Step-4, set OuterParams Reduction dim: bdimy.
+  op.bdimy = threads_per_block / op.bdimx;
+
+  // Step-5, special case, when inner_dim_numel <= 1024, bdimx is usually small
+  // after divide by vect and batch. In this case, bdimy is used to parallelize
+  // outer_dim instead of inner_dim. This pattern is named multi reductions per
+  // block (mrpb).
+  if (inner_dim_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+    rparams->tidx_for_outer_reduction = true;
+    constexpr int64_t threads_per_block_mrpb = 512;
+
+    // Step-1, InnerParams, Reduction dim: vect(reuse), batch, bdimx
+    ip.batch = 1;
+    ip.bdimx = inner_dim_numel / ip.vect;
+
+    // Step-2, InnerParams, Iteration dim: gdimy, bdimy (same as op.bdimy)
+    reg_per_thread = getEstimatedRegisterUsage(ip.vect * ip.batch);
+    threads_per_sm = getThreadsPerSMGivenRegPerThread(reg_per_thread);
+    blocks_per_sm = getBlocksPerSM(
+        threads_per_sm, threads_per_block_mrpb, dev_prop->warpSize);
+    ip.gdimy = blocks_per_sm * device_multiprocessor_count;
+
+    // Step-3, OuterParams, Iteration dim: unroll_factor, bdimy, gdimy(same as
+    // ip.gdimy)
+    op.gdimy = ip.gdimy;
+    op.unroll_factor = std::min(ip.vect, (int64_t)2);
```

Comment:
Curious why the same value is used for both ip.vect and op.unroll_factor in the small case (with min(2)). Isn't this scheduler only enabled when some factor of vectorization is possible? If so, `vect` should be >=2, so isn't this just setting unroll_factor by 2?

Reply:
OK, can you please leave a comment why it's 2?

Reply:
added